staging/src/k8s.io/dynamic-resource-allocation/structured/celcache.go

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package structured
+
+import (
+	"sync"
+
+	"github.com/golang/groupcache/lru"
+
+	"k8s.io/dynamic-resource-allocation/cel"
+)
+
+// CELCache is an LRU cache for a compiled CEL expression.
+type CELCache struct {
+	mutex sync.Mutex
+	cache *lru.Cache
+}
+
+// NewCELCache returns a CELCache
+func NewCELCache(maxCacheEntries int) *CELCache {
+	return &CELCache{
+		cache: lru.New(maxCacheEntries),
+	}
+}
+
+func (c *CELCache) add(expression string, expr *cel.CompilationResult) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.cache.Add(expression, expr)
+}
+
+func (c *CELCache) get(expression string) *cel.CompilationResult {
+	c.mutex.Lock()
```

Comment:
If sync.RWMutex is used then we could use `c.mutex.RLock()` and `c.mutex.RUnlock()`